2-Draft-Accepted/RFCNNNN-Support-Experimental-Features.md

```diff
@@ -0,0 +1,626 @@
+
+When reading the experimental feature list from configuration file,
+shall we parse the dependencies at the same time or defer the dependency resolution until module loading?
+
+As described in the [`Experimental Feature Dependencies`](#experimental-feature-dependencies) section,
+an experimental feature may have a chain dependency on other experimental features.
+In that section, the proposal is to check for dependencies at module loading time.
+
+Another proposal is to resolve the dependencies when loading the configuration file,
+so that PowerShell can fail early when dependencies are not satisfied.
+The fail-early idea is desired,
+however, PowerShell loads the enabled experimental feature list at very early stage,
+where the engine is not ready for the operations required for resolving dependencies,
+such as module discovery.
+Also, there will be an impact to the startup time if we do this resolution when loading the configuration file.
+Lastly, the module that exposes an experimental feature may not be in the module path yet
+when PowerShell starts,
+in which case it's impossible to resolve the experimental feature name.
+
+### How to override the instance configuration file -- rewrite command-line parsing vs. use environment variable
+
+A new command-line flag `-settingsFile` was added to `pwsh` to allow
+a user to override the instance configuration file `$PSHOME\powershell.config.json` using another configuration file.
+This flag is very useful in testing the experimental features.
+
+However, there are some amount of work that occurs before the command-line parameter parsing,
+so if that amount of work has experimental features,
+it would be hard to flag them in testing using `-settingsFile`.
+This problem is seen on Linux with logging configuration due to
+the logging happens at a very early stage when PowerShell starts.
+
+There are two proposals here for solving this problem:
+
+1. Move the command-line parameter parsing earlier in the startup call-path.
+   The main parsing logic depends on other components like the `InitialSessionState`,
+   so it cannot be moved to a very early stage.
+   A possible solution is to parse the command-line parameters in two passes.
+   The first pass is to convert the arguments into a more structured format,
+   and process flags like `-settingsFile` at a very early stage.
+   The second pass is the main parsing logic.
+2. Use an environment variable `PSSettingsFile` to point to the configuration file to use.
+   By using the environment variable,
+   overriding the instance configuration file can happen as early as the singleton instance of `PowerShellConfig` gets created,
+   which would be the first time PowerShell attempts to read the configuration file.
+
+The first proposal will cost way more than the second proposal.
+Since the configuration file overriding functionality is mainly for testing,
+the second proposal may be more efficient cost-wise.
```

Comment:
I believe it makes sense to split on early parsing and later appling of command-line parameters.

Reply:
I tend to agree but it may not be justified here.